src/java.base/share/classes/jdk/internal/classfile/impl/Util.java

```diff
@@ -84,9 +84,26 @@ public static int maxLocals(int flags, MethodTypeDesc mDesc) {
         return count;
     }
 
-    public static String toClassString(String desc) {
-        //TODO: this doesn't look right L ... ;
-        return desc.replace('/', '.');
+    /**
+     * Convert a descriptor of classes or interfaces or arrays, or an internal
+     * name of a class or interface, into a fully qualified binary name, that can
+     * be resolved by {@link Class#forName(String) Class::forName}. Primitive type
+     * descriptors should never be passed into this method.
+     *
+     * @param descOrInternalName a descriptor or internal name
+     * @return the fully qualified binary name
+     */
+    public static String toBinaryName(String descOrInternalName) {
+        if (descOrInternalName.startsWith("L")) {
+            // descriptors of classes or interfaces
+            if (descOrInternalName.length() <= 2 || !descOrInternalName.endsWith(";")) {
+                throw new IllegalArgumentException(descOrInternalName);
+            }
+            return descOrInternalName.substring(1, descOrInternalName.length() - 1).replace('/', '.');
+        } else {
```

Comment:
Note that this won’t work right if this method is called with the internal name of a class whose FQCN starts with `L`, the correct implementation would be:
```suggestion
        if (descOrInternalName.startsWith("L") && descOrInternalName.endsWith(";")) {
            // descriptors of classes or interfaces
            if (descOrInternalName.length() <= 2) {
                throw new IllegalArgumentException(descOrInternalName);
            }
            return descOrInternalName.substring(1, descOrInternalName.length() - 1).replace('/', '.');
        } else {
```

Reply:
Honestly, this API should be changed to take a ClassDesc instead.

Reply:
In fact, the specification (javadoc) on this method is wrong: it will never take an internal name; it only takes descriptors and tries to simulate the result of `Class::getName`, since `ClassDesc` lacks such functionality.